README.md
# angular-error-tooltip

Lightweight tooltip directive that displays a FormControl's error messages to user.
<!--![angular 2 tooltip](https://raw.githubusercontent.com/NaughtyMC/angular-error-tooltip/master/resources/tooltip-example.png)-->

## Installation

1. Install npm module:

    `npm install angular-error-tooltip --save`

2. If you are using system.js you may want to add this into `map` and `package` config:

    ```json
    {
        "map": {
            "angular-error-tooltip": "node_modules/angular-error-tooltip"
        },
        "packages": {
            "angular-error-tooltip": { "main": "index.js", "defaultExtension": "js" }
        }
    }
    ```

## Samples

[Sample 1](https://github.com/NaughtyMC/angular-error-tooltip/blob/master/sample/sample1):

![sample1](sample/demo/demo.gif "sample1") 


## Contributing
Please read [CONTRIBUTING.md](CONTRIBUTING.md) for details on our code of conduct.

## License
This project is licensed under the MIT License. See the [LICENSE.txt](LICENSE.txt) file for details.

## Acknowledgements
 * Special thanks to [Umed Khudoiberdiev](https://github/com/pleerock), whose [ngx-tooltip](https://github.com/pleerock/ng2-tooltip) served as a backbone for this project


CHANGELOG.md
# [1.0.0](https://github.com/NaughtyMC/angular-error-tooltip/compare/1.0.0-beta.2...v1.0.0) (2017-02-22)
* Adds samples to project and README.md ([#1](https://github.com/NaughtyMC/angular-error-tooltip/issues/2))

# [1.0.0-beta.2](https://github.com/NaughtyMC/angular-error-tooltip/compare/v1.0.0-beta.1...1.0.0-beta.2) (2017-02-19)
* Outline and error tooltip do not appear when page is loaded ([#2](https://github.com/NaughtyMC/angular-error-tooltip/commit/c7fa0fac3eacb236e197193df5257f88c896fadc))
* Error tooltip toggles visibility as the control gains/loses focus

# 1.0.0-beta.1 (2017-02-14)
* Implements core functionality
* When FormControl fails validation, **angular-error-message** dynamically creates a tooltip with applicable error message


src/ErrorTooltip.ts
import { Directive, HostListener, HostBinding, ComponentRef, ViewContainerRef, Input, ComponentFactoryResolver, ComponentFactory, OnInit, OnDestroy, Optional, Host, SkipSelf } from "@angular/core";
import { NgControl, FormGroupDirective } from '@angular/forms'
import { Subscription } from 'rxjs/Subscription';
import { ErrorTooltipContent } from "./ErrorTooltipContent";

@Directive({
    selector: "[errorTooltip]"
})
export class ErrorTooltip implements OnInit, OnDestroy {

    // -------------------------------------------------------------------------
    // Properties
    // -------------------------------------------------------------------------

    private tooltip: ComponentRef<ErrorTooltipContent>;
    private tooltipFactory: ComponentFactory<ErrorTooltipContent>;
    private initialized: boolean;
    private statusSubscription: Subscription;
    private formSubmissionSubscription: Subscription;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    constructor
        (private viewContainerRef: ViewContainerRef,
        private resolver: ComponentFactoryResolver,
        private control: NgControl,
        @Optional() @Host() @SkipSelf() private parent: FormGroupDirective
    ) {
    }

    // -------------------------------------------------------------------------
    // Inputs / Outputs
    // -------------------------------------------------------------------------

    @Input()
    tooltipAnimation: boolean = true;

    @Input()
    tooltipPlacement: "top"|"bottom"|"left"|"right" = "bottom";

    // -------------------------------------------------------------------------
    // Methods
    // -------------------------------------------------------------------------
    ngOnInit() {
        this.tooltipFactory = this.resolver.resolveComponentFactory(ErrorTooltipContent);

        this.statusSubscription = this.control.statusChanges.subscribe(
            status => this.processStatus(status)
        );

        this.formSubmissionSubscription = this.parent.ngSubmit.subscribe((e: any) => this.processStatus(this.control.control.status as ('VALID' | 'INVALID' | 'PENDING' | 'DISABLED')));
    }

    @HostListener('focusin')
    show(): void {
        this.tooltip && this.tooltip.instance.show();
    }
    @HostListener('focusout')
    hide(): void {
        this.tooltip && this.tooltip.instance.hide();
    }
    ngOnDestroy() {
        this.statusSubscription && this.statusSubscription.unsubscribe();
        this.formSubmissionSubscription && this.formSubmissionSubscription.unsubscribe();
    }

    private processStatus = (status: 'VALID' | 'INVALID' | 'PENDING' | 'DISABLED'): void => {
        switch (status) {
            case 'INVALID':
                this.update();
                break;
            case 'VALID':
            case 'DISABLED':
                this.destroy();
                break;
            case 'PENDING':
            default:
                break; // do nothing
        }
    }

    private create = (): void => {
        if (!this.tooltip && this.control.invalid) {

            this.tooltip = this.viewContainerRef.createComponent(this.tooltipFactory);
            this.tooltip.instance.initialize(this.control, this.viewContainerRef.element.nativeElement, this.tooltipPlacement);
        }
    }
    private update = (): void => {
        if (this.tooltip) {
            this.tooltip.instance.updateErrorMessage();
        } else {
            this.create();
        }
    }
    private destroy = (): void => {
        if (this.tooltip) {
            this.tooltip.destroy();
            this.tooltip = null;
        }    
    }

    @HostBinding('style.border') get hostBorder() { return this.tooltip ? '2px solid #FF4844' : null };
}

src/TooltipContent.ts
import {Component, Input, AfterViewInit, ElementRef, ChangeDetectorRef} from "@angular/core";

@Component({
    selector: "tooltip-content",
    template: `
<div class="tooltip {{ placement }}"
     [style.top]="top + 'px'"
     [style.left]="left + 'px'"
     [class.in]="isIn"
     [class.fade]="isFade"
     role="tooltip">
    <div class="tooltip-arrow"></div> 
    <div class="tooltip-inner">
        <ng-content></ng-content>
        {{ content }}
    </div> 
</div>
`,
    styles: [`
.tooltip-inner {
    color: #A94442;
    border: 1px solid #000;
    background-color: #F2DEDE;
    font-size: 14px;
}
`]
})
export class TooltipContent implements AfterViewInit {

    // -------------------------------------------------------------------------
    // Inputs / Outputs 
    // -------------------------------------------------------------------------

    @Input()
    hostElement: HTMLElement;

    @Input()
    content: string;

    @Input()
    placement: "top"|"bottom"|"left"|"right" = "bottom";

    @Input()
    animation: boolean = true;

    // -------------------------------------------------------------------------
    // Properties
    // -------------------------------------------------------------------------

    top: number = -100000;
    left: number = -100000;
    isIn: boolean = false;
    isFade: boolean = false;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    constructor(private element: ElementRef,
                private cdr: ChangeDetectorRef) {
    }

    // -------------------------------------------------------------------------
    // Lifecycle callbacks
    // -------------------------------------------------------------------------

    ngAfterViewInit(): void {
        this.show();
        this.cdr.detectChanges();
    }

    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    
    show(): void {
        if (!this.hostElement)
            return;

        const p = this.positionElements(this.hostElement, this.element.nativeElement.children[0], this.placement);
        this.top = p.top;
        this.left = p.left;
        this.isIn = true;
        if (this.animation)
            this.isFade = true;
    }

    hide(): void {
        this.top = -100000;
        this.left = -100000;
        this.isIn = true;
        if (this.animation)
            this.isFade = false;
    }

    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------

    private positionElements(hostEl: HTMLElement, targetEl: HTMLElement, positionStr: string, appendToBody: boolean = false): { top: number, left: number } {
        let positionStrParts = positionStr.split("-");
        let pos0 = positionStrParts[0];
        let pos1 = positionStrParts[1] || "center";
        let hostElPos = appendToBody ? this.offset(hostEl) : this.position(hostEl);
        let targetElWidth = targetEl.offsetWidth;
        let targetElHeight = targetEl.offsetHeight;
        let shiftWidth: any = {
            center: function (): number {
                return hostElPos.left + hostElPos.width / 2 - targetElWidth / 2;
            },
            left: function (): number {
                return hostElPos.left;
            },
            right: function (): number {
                return hostElPos.left + hostElPos.width;
            }
        };

        let shiftHeight: any = {
            center: function (): number {
                return hostElPos.top + hostElPos.height / 2 - targetElHeight / 2;
            },
            top: function (): number {
                return hostElPos.top;
            },
            bottom: function (): number {
                return hostElPos.top + hostElPos.height;
            }
        };

        let targetElPos: { top: number, left: number };
        switch (pos0) {
            case "right":
                targetElPos = {
                    top: shiftHeight[pos1](),
                    left: shiftWidth[pos0]()
                };
                break;
            
            case "left":
                targetElPos = {
                    top: shiftHeight[pos1](),
                    left: hostElPos.left - targetElWidth
                };
                break;
            
            case "bottom":
                targetElPos = {
                    top: shiftHeight[pos0](),
                    left: shiftWidth[pos1]()
                };
                break;
            
            default:
                targetElPos = {
                    top: hostElPos.top - targetElHeight,
                    left: shiftWidth[pos1]()
                };
                break;
        }

        return targetElPos;
    }

    private position(nativeEl: HTMLElement): { width: number, height: number, top: number, left: number } {
        let offsetParentBCR = { top: 0, left: 0 };
        const elBCR = this.offset(nativeEl);
        const offsetParentEl = this.parentOffsetEl(nativeEl);
        if (offsetParentEl !== window.document) {
            offsetParentBCR = this.offset(offsetParentEl);
            offsetParentBCR.top += offsetParentEl.clientTop - offsetParentEl.scrollTop;
            offsetParentBCR.left += offsetParentEl.clientLeft - offsetParentEl.scrollLeft;
        }

        const boundingClientRect = nativeEl.getBoundingClientRect();
        return {
            width: boundingClientRect.width || nativeEl.offsetWidth,
            height: boundingClientRect.height || nativeEl.offsetHeight,
            top: elBCR.top - offsetParentBCR.top,
            left: elBCR.left - offsetParentBCR.left
        };
    }

    private offset(nativeEl:any): { width: number, height: number, top: number, left: number } {
        const boundingClientRect = nativeEl.getBoundingClientRect();
        return {
            width: boundingClientRect.width || nativeEl.offsetWidth,
            height: boundingClientRect.height || nativeEl.offsetHeight,
            top: boundingClientRect.top + (window.pageYOffset || window.document.documentElement.scrollTop),
            left: boundingClientRect.left + (window.pageXOffset || window.document.documentElement.scrollLeft)
        };
    }

    private getStyle(nativeEl: HTMLElement, cssProp: string): string {
        if ((nativeEl as any).currentStyle) // IE
            return (nativeEl as any).currentStyle[cssProp];

        if (window.getComputedStyle)
            return (window.getComputedStyle(nativeEl) as any)[cssProp];
        
        // finally try and get inline style
        return (nativeEl.style as any)[cssProp];
    }

    private isStaticPositioned(nativeEl: HTMLElement): boolean {
        return (this.getStyle(nativeEl, "position") || "static" ) === "static";
    }

    private parentOffsetEl(nativeEl: HTMLElement): any {
        let offsetParent: any = nativeEl.offsetParent || window.document;
        while (offsetParent && offsetParent !== window.document && this.isStaticPositioned(offsetParent)) {
            offsetParent = offsetParent.offsetParent;
        }
        return offsetParent || window.document;
    }

}

sample/sample1/main.ts
import { Component, NgModule } from "@angular/core";
import { platformBrowserDynamic } from "@angular/platform-browser-dynamic";
import { BrowserModule } from "@angular/platform-browser";
import { ReactiveFormsModule } from '@angular/forms';
import { ErrorTooltipModule } from "../../src/index";

@Component({
    selector: "app",
    template: `
<div class="container">
    <form novalidate [formGroup]="sampleForm" class="form-horizontal" role="form">
        <div class="form-group">
            <label class="control-label col-sm-3" for="sample1">Required:</label>
            <div class="col-sm-9">
                <input type="text" class="form-control" formControlName="sample1" errorTooltip tooltipPlacement="top">
            </div>
        </div>
        <div class="form-group">
            <label class="control-label col-sm-3" for="sample2">Minimum length 3:</label>
            <div class="col-sm-9">
                <input type="text" class="form-control" formControlName="sample2" errorTooltip tooltipPlacement="top">
            </div>
        </div>
        <div class="form-group">
            <label class="control-label col-sm-3" for="sample3">Maximum length 3:</label>
            <div class="col-sm-9">
                <input type="text" class="form-control" formControlName="sample3" errorTooltip tooltipPlacement="top">
            </div>
        </div>
        <div class="col-sm-2 pull-right">
            <button type="submit" class="btn btn-primary btn-lg btn-block">SUBMIT</button>
        </div>
    </form>
</div>
`
})
export class Sample1App {

}

@NgModule({
    imports: [
        BrowserModule,
        ErrorTooltipModule
    ],
    declarations: [
        Sample1App
    ],
    bootstrap: [
        Sample1App
    ]
})
export class Sample1Module {

}

platformBrowserDynamic().bootstrapModule(Sample1Module);

src/ErrorTooltipContent.ts
import { Component, Input, ElementRef, ChangeDetectorRef } from "@angular/core";
import { NgControl } from '@angular/forms';

@Component({
    selector: "error-tooltip-content",
    template: `
<div class="tooltip {{ placement }}"
     [style.top.px]="tooltipPosition.top"
     [style.left.px]="tooltipPosition.left"
     [class.in]="isIn"
     [class.fade]="isFade"
     role="tooltip">
    <div class="tooltip-arrow"></div> 
    <div class="tooltip-inner">
        <ng-content></ng-content>
        {{ errorMessage }}
    </div> 
</div>
`,
    styles: [`
.tooltip-inner {
    color: #A94442;
    border: 1px solid #000;
    background-color: #F2DEDE;
    font-size: 14px;
}
`]
})
export class ErrorTooltipContent {

    // -------------------------------------------------------------------------
    // Properties
    // -------------------------------------------------------------------------

    hostElement: HTMLElement;
    control: NgControl;
    placement: "top" | "bottom" | "left" | "right" = "bottom";

    private get errorMessage(): string { return this.errorMessages ? this.errorMessages.join('\n') : null; }
    private errorMessages: string[];

    tooltipPosition: { top: number, left: number } = { top: -100000, left: -100000 };
    isIn: boolean = true;
    isFade: boolean = true;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    constructor(private element: ElementRef,
        private cdr: ChangeDetectorRef) {
    }

    // -------------------------------------------------------------------------
    // Lifecycle callbacks
    // -------------------------------------------------------------------------
    

    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------

    show = (): void => {
        if (!this.hostElement)
            return;

        this.positionTooltip();
        this.isIn = true;
        this.isFade = true;
    }

    hide = (): void => {
        this.tooltipPosition = { top: -100000, left: -100000 };
        this.isIn = true;
        this.isFade = false;
    }

    initialize = (control: NgControl, hostElement: HTMLElement, placement: 'top' | 'bottom' | 'left' | 'right'): void => {
        this.control = control;
        this.hostElement = hostElement;
        this.placement = placement;

        this.updateErrorMessage();
        this.cdr.detectChanges();
    }

    updateErrorMessage = (): void => {
        let errMsgs: string[] = [];
        for (let propName in this.control.errors) {
            if (this.control.errors.hasOwnProperty(propName)) {

                let errMsg: string;
                switch (propName) {
                    case 'required':
                        errMsg = 'Required';
                        break;
                    case 'minlength':
                        errMsg = `Must contain at least ${this.control.errors[propName].requiredLength} characters`
                        break;
                    case 'maxlength':
                        errMsg = `Cannot contain more than ${this.control.errors[propName].requiredLength} characters`
                        break;
                    case 'pattern':
                        errMsg = 'Invalid value';
                        break;
                    default:
                        break; // Do nothing
                }

                if (errMsg) errMsgs.push(errMsg);
            }
        }

        let updatePositioning = this.hasNewErrorMessages(errMsgs);
        this.errorMessages = errMsgs;

        if (updatePositioning) {
            this.cdr.detectChanges();
            this.positionTooltip(); // adjust positioning
        }
    }

    // -------------------------------------------------------------------------
    // Private Methods
    // -------------------------------------------------------------------------

    private positionElements(hostEl: HTMLElement, targetEl: HTMLElement, positionStr: string, appendToBody: boolean = false): { top: number, left: number } {
        let positionStrParts = positionStr.split("-");
        let pos0 = positionStrParts[0];
        let pos1 = positionStrParts[1] || "center";
        let hostElPos = appendToBody ? this.offset(hostEl) : this.position(hostEl);
        let targetElWidth = targetEl.offsetWidth;
        let targetElHeight = targetEl.offsetHeight;
        let shiftWidth: any = {
            center: function (): number {
                return hostElPos.left + hostElPos.width / 2 - targetElWidth / 2;
            },
            left: function (): number {
                return hostElPos.left;
            },
            right: function (): number {
                return hostElPos.left + hostElPos.width;
            }
        };

        let shiftHeight: any = {
            center: function (): number {
                return hostElPos.top + hostElPos.height / 2 - targetElHeight / 2;
            },
            top: function (): number {
                return hostElPos.top;
            },
            bottom: function (): number {
                return hostElPos.top + hostElPos.height;
            }
        };

        let targetElPos: { top: number, left: number };
        switch (pos0) {
            case "right":
                targetElPos = {
                    top: shiftHeight[pos1](),
                    left: shiftWidth[pos0]()
                };
                break;

            case "left":
                targetElPos = {
                    top: shiftHeight[pos1](),
                    left: hostElPos.left - targetElWidth
                };
                break;

            case "bottom":
                targetElPos = {
                    top: shiftHeight[pos0](),
                    left: shiftWidth[pos1]()
                };
                break;

            default:
                targetElPos = {
                    top: hostElPos.top - targetElHeight,
                    left: shiftWidth[pos1]()
                };
                break;
        }

        return targetElPos;
    }

    private position(nativeEl: HTMLElement): { width: number, height: number, top: number, left: number } {
        let offsetParentBCR = { top: 0, left: 0 };
        const elBCR = this.offset(nativeEl);
        const offsetParentEl = this.parentOffsetEl(nativeEl);
        if (offsetParentEl !== window.document) {
            offsetParentBCR = this.offset(offsetParentEl);
            offsetParentBCR.top += offsetParentEl.clientTop - offsetParentEl.scrollTop;
            offsetParentBCR.left += offsetParentEl.clientLeft - offsetParentEl.scrollLeft;
        }

        const boundingClientRect = nativeEl.getBoundingClientRect();
        return {
            width: boundingClientRect.width || nativeEl.offsetWidth,
            height: boundingClientRect.height || nativeEl.offsetHeight,
            top: elBCR.top - offsetParentBCR.top,
            left: elBCR.left - offsetParentBCR.left
        };
    }

    private offset(nativeEl: any): { width: number, height: number, top: number, left: number } {
        const boundingClientRect = nativeEl.getBoundingClientRect();
        return {
            width: boundingClientRect.width || nativeEl.offsetWidth,
            height: boundingClientRect.height || nativeEl.offsetHeight,
            top: boundingClientRect.top + (window.pageYOffset || window.document.documentElement.scrollTop),
            left: boundingClientRect.left + (window.pageXOffset || window.document.documentElement.scrollLeft)
        };
    }

    private getStyle(nativeEl: HTMLElement, cssProp: string): string {
        if ((nativeEl as any).currentStyle) // IE
            return (nativeEl as any).currentStyle[cssProp];

        if (window.getComputedStyle)
            return (window.getComputedStyle(nativeEl) as any)[cssProp];

        // finally try and get inline style
        return (nativeEl.style as any)[cssProp];
    }

    private isStaticPositioned(nativeEl: HTMLElement): boolean {
        return (this.getStyle(nativeEl, "position") || "static") === "static";
    }

    private parentOffsetEl(nativeEl: HTMLElement): any {
        let offsetParent: any = nativeEl.offsetParent || window.document;
        while (offsetParent && offsetParent !== window.document && this.isStaticPositioned(offsetParent)) {
            offsetParent = offsetParent.offsetParent;
        }
        return offsetParent || window.document;
    }

    private positionTooltip = (): void => {
        this.tooltipPosition = this.positionElements(this.hostElement, this.element.nativeElement.children[0], this.placement);
    }

    private hasNewErrorMessages = (newErrorMessages: string[]): boolean => {
        if (!this.errorMessages || this.errorMessages.length !== newErrorMessages.length) {
            return true;
        } else {
            for (let i = 0; i < newErrorMessages.length; i++) {
                if (this.errorMessages.findIndex((errorMessage) => { return errorMessage === newErrorMessages[i]; }) === -1) {
                    return true;
                }
            }

            return false;
        }
    }
}